DR-479 soft deletes api

thathert opened the pull request:
This implements the soft delete by file batch API. Needs more tests since it only has happy path right now, but should be ready for initial review.

Commit:
add data deletion endpoint and models to openapi spec

Commit:
add required fields

Commit:
WIP

Commit:
progress on the model

Commit:
missed rename

Commit:
intermediate progress from pairing

Commit:
creating external tables

Commit:
temporarily removing locks

Commit:
happy path working

Commit:
clean up + finish happy path test

ddietterich commented on src/main/resources/data-repository-openapi.yaml:
```diff
@@ -2658,6 +2692,56 @@ definitions:
+          - csv
+      path:
+        type: string
+        description: a gs://path/to/a/file that can use wildcards(*)
```
What is the use case for wildcards here? Not having an explicit file seems a bit hazardous to me.

danxmoran replied:
Monster is the use-case. We're running BigQuery jobs outside of the repo to compute our row-level diffs, and the results pop out as BigQuery tables. We have to export back to GCS for the repo APIs to work. If the exported payload is "too big", BQ will refuse to export it into a single file, so we end up dealing with part-files again.

danxmoran replied:
Docs: https://cloud.google.com/bigquery/docs/exporting-data#exporting_data_into_one_or_more_files

danxmoran replied:
And for one more data point: The size limit for a single file is 1 GiB. If I run "select datarepo_row_id" from one of the V2F tables in prod, BigQuery shows me that it'll scan 48 GiB of data. We don't intend to soft-delete in V2F any time soon, but I think the size limit is a relatively tight constraint for these big datasets.

And of course the issue is that once we let BigQuery shard, it does whatever it feels like doing, and we go from dealing with 1 consistent pathspec to `N` dynamic pathspecs

ddietterich replied:
Thanks, I cut'n'pasted your reply into the design-ish document.

ddietterich commented on src/main/resources/data-repository-openapi.yaml:
```diff
@@ -2658,6 +2692,56 @@ definitions:
+      deleteType:
+        type: string
+        enum:
+        - soft
```
I like this structure. We know we will have to extend it, so nice to have the layout.

ddietterich commented on src/main/java/bio/terra/service/dataset/DataDeletionRequestValidator.java:
```diff
@@ -0,0 +1,35 @@
+
+    private void validateFileSpec(DataDeletionTableModel fileSpec, Errors errors) {
+        String tableName = fileSpec.getTableName();
+        if (tableName == null || tableName.equals("")) {
```
You can do this check with `StringUtils.isEmpty()` or if you want to know if it contains spaces, `isBlank`.

ddietterich commented on src/main/java/bio/terra/service/dataset/exception/InvalidLockUsageException.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.List;
+
+public class InvalidLockUsageException extends InternalServerErrorException {
```
nit: this seems mis-named to me. It is more of an IllegalArgument exception - the lock/unlock step was not properly constructed.

From the name, I thought it would have something to do with locking.

thathert replied:
Good point, will update.

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/LockDatasetBaseStep.java:
```diff
@@ -0,0 +1,52 @@
+
+import java.util.UUID;
+
+// I was very tempted to call this LockStep...
```
I'm not understanding what this class adds. It looks like it is of no use to the UnlockStep. So it doesn't make sense to me to make it a base class.

thathert replied:
It is to share the logic for fetching a dataset name in the case where only a dataset id is available in the input params. Feels like it could also be on a Utils class but I want to avoid having too many of those. What do you think?

ddietterich replied:
It seems odd to have a base class that is only of use to one of its subclasses. Unless there is another place to use it, I'd just move the code inside the lock step. Or move the `getDatasetName` method to Utils if it is of use anywhere else.

ddietterich replied:
The root cause of this problem is that the datasetDao lock/unlock calls want a dataset name. Those calls are only used in the case where the dataset exists already, so they should take `datasetId` instead.

SnapshotDao has the same problem. I wish I had noticed that on those reviews. I really think we should not be using names for these lock calls. We could either:
1. Add that change to this PR
1. Make a new PR that just changes from name to id, and get that in ahead of this PR

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/CreateExternalTablesStep.java:
```diff
@@ -0,0 +1,110 @@
+
+    private void validateFilesExistForTables(DataDeletionRequest request) {
+        // TODO: revisit the name of this? table could be confusing here
+        List<DataDeletionTableModel> tablesMissingFiles = request.getTables()
```
minor: You could do it in one pass with an explicit loop, building the bad array as you went.

thathert replied:
Done

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/CreateExternalTablesStep.java:
```diff
@@ -0,0 +1,110 @@
+            .get(JobMapKeys.REQUEST.getKeyName(), DataDeletionRequest.class);
+    }
+
+    private Dataset dataset(FlightContext context) {
```
These method names look more like setters than getters. Maybe prefix with `get`?

thathert replied:
Will do.

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/CreateExternalTablesStep.java:
```diff
@@ -0,0 +1,110 @@
+        String suffix = suffix(context);
+        DataDeletionRequest dataDeletionRequest = request(context);
+
+        validateFilesExistForTables(dataDeletionRequest);
```
Where do you detect that a table named in the request does not exist in the dataset?

thathert replied:
That needs to happen in the flight -- it's not currently happening.

- [x] check that tables exist in the dataset
- [x] validate row ids against the dataset before soft deleting

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DatasetDataDeleteFlight.java:
```diff
@@ -0,0 +1,54 @@
+        // need to lock, need dataset name and flight id
+        addStep(new LockDatasetStep(datasetDao));
+
+        // check to see access to file and create external temp table
```
nit: comment is singular, but I believe these operate over multiple files and tables.

thathert replied:
Yep, good point.

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DatasetDataDeleteFlight.java:
```diff
@@ -0,0 +1,54 @@
+        addStep(new UnlockDatasetStep(datasetDao));
+
+        // cleanup
+        addStep(new DropExternalTablesStep(bigQueryPdao, datasetService));
```
A second run of soft deletes will collide on the names of external tables, if it starts before this completes.
I think either:
- Randomize the external table names in some way so they won't collide
- Reverse the order of this and the unlock.

thathert replied:
Soft deletes use the flight id for the suffix of the external tables. This should prevent collisions, right?

ddietterich replied:
Yes, that'll do it.

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DataDeletionStep.java:
```diff
@@ -0,0 +1,86 @@
+            .get(JobMapKeys.REQUEST.getKeyName(), DataDeletionRequest.class);
+    }
+
+    private Dataset dataset(FlightContext context) {
```
Ditto

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DropExternalTablesStep.java:
```diff
@@ -0,0 +1,62 @@
+            .get(JobMapKeys.REQUEST.getKeyName(), DataDeletionRequest.class);
+    }
+
+    private Dataset dataset(FlightContext context) {
```
Same comment re: the method names

ddietterich replied:
You didn't change the method names. Did you intend to?

danxmoran commented on src/main/java/bio/terra/service/dataset/DataDeletionRequestValidator.java:
```diff
@@ -0,0 +1,35 @@
+            errors.rejectValue("tables.tableName", "TableNameMissing",
+                "Requires a table name");
+        }
+    }
```
BigQuery supports at most 1 wildcard `*` in GCS references, would it be worth adding that check here?

thathert replied:
Yes, I want to re-use the logic that we already have for gcs-path-with-wildcard validation.

danxmoran commented on src/main/resources/data-repository-openapi.yaml:
```diff
@@ -2658,6 +2692,56 @@ definitions:
         items:
           $ref: '#/definitions/PolicyModel'
 
+  DataDeletionRequest:
```
Should this be named more specifically to mention tables? Or will file deletions eventually shift over to use this model, too?

thathert replied:
I imagine the file model will be different, updating.

ddietterich commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1053,53 @@ private LegacySQLTypeName translateType(String datatype) {
+     * @param tableNames list of table names that should have corresponding external tables with row ids to soft delete
+     * @param suffix a bq-safe version of the flight id to prevent different flights from stepping on each other
+     */
+    public TableResult applySoftDeletes(Dataset dataset,
```
Should we make sure that the row ids actually exist in the base table?
Should we avoid duplicates in the soft delete table?

thathert replied:
Yep was thinking about that.

thathert replied:
Added a checklist item about this

danxmoran commented on src/main/resources/data-repository-openapi.yaml:
```diff
@@ -2658,6 +2692,56 @@ definitions:
+          - csv
+      path:
+        type: string
+        description: a gs://path/to/a/file that can use wildcards(*)
```
```suggestion
        description: a gs://path/to/a/file that can include a wildcard (*)
```

danxmoran commented on src/main/java/bio/terra/service/dataset/DatasetService.java:
```diff
@@ -133,4 +135,15 @@ public String removeDatasetAssetSpecifications(
             .submit();
     }
+
+    public String deleteDatasetData(
```
Similar question as on the API, should this name be more specific since it's only dealing with tables?

ddietterich reviewed (changes requested):
My main concerns are that it looks like we are not checking that incoming row ids exist in the base table or that they might already exist in the soft deletes table.

Other minor comments

danxmoran commented on src/main/java/bio/terra/service/dataset/flight/datadelete/CreateExternalTablesStep.java:
```diff
@@ -0,0 +1,110 @@
+        // TODO: support other file types (assuming header-less csv here)
+        String path  = table.getGcsFileSpec().getPath();
+        try {
+            return GcsPdao.getBlobFromGsPath(storage, path).exists();
```
This won't work if there's a wildcard in the path. When I added wildcard support on the ingest side, we [stripped away](https://github.com/DataBiosphere/jade-data-repo/pull/315/files#diff-536f43809cfb93169e496e7813e081daL60-R55) most up-front validation, and [caught the "not found" case](https://github.com/DataBiosphere/jade-data-repo/pull/315/files#diff-bdcc1fe50c17d33d41a93d142e860a12R355-R357) after it was raised by BigQuery.

danxmoran commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1053,53 @@ private LegacySQLTypeName translateType(String datatype) {
+        return String.format("%s%s_%s", PDAO_EXTERNAL_TABLE_PREFIX, tableName, suffix);
+    }
+
+    public String createExternalTable(Dataset dataset, String path, String tableName, String suffix) {
```
From the flight code I see that both `path` and `tableName` are derived from the same `table` object. Could we refactor this to pass that `table` in here? It'd make the flight code simpler, and reduce the chances of somebody passing in mismatched args.

danxmoran replied:
Also, on naming: This is named really generically, but hard-coded to a specific use-case. Should we either:
* Rename the method to reflect its specific use, or
* Add more parameters to the method for the `Schema` and `FormatOptions`, and let the flight be the one to do the hard-coding for now?

thathert replied:
Good point, will go with the second option.

thathert replied:
Changed my mind and went with first. I prefer to avoid "leaking" BQ objects out into the steps.. it seems like we get better encapsulation having the PDAO be responsible for handling things like organizing names and dealing with `Schema` creation, etc.

danxmoran commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1053,53 @@ private LegacySQLTypeName translateType(String datatype) {
+     */
+    public TableResult applySoftDeletes(Dataset dataset,
+                                 List<String> tableNames,
+                                 Map<String, String> softDeleteTableNameLookup,
```
Same as above, I think this signature & the calling code would become simpler and more difficult to call incorrectly if this took a list of table objects, and called the "right" getters for each. It's the pattern used elsewhere in the PDAO.

thathert replied:
Done

danxmoran commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1053,53 @@ private LegacySQLTypeName translateType(String datatype) {
         }
     }
+
+    private String externalTableName(String tableName, String suffix) {
```
We've mostly banished the magic string prefixes from the PDAO, though I see the need for something like this one. Instead of having every method below take both a `tableName` and a `suffix`, immediately passing both into this method, could we push the computation into a setup step in the flight? Then the flight context can track the full external table name instead of the suffix.

thathert replied:
See my comment above. I prefer to have the flights only care about the suffix, and not have to remember things that are more the responsibility of the PDAO, like how to use the right prefix.

ddietterich replied:
@danxmoran Not sure what you mean "banished". We need to use name prefixes to avoid collisions with the things we let the users name.

@thathert I don't mind this as it is. If you _really_ want to encapsulate the name, you could make a step that calls PDAO to generate external table names. They could just be in the form `<pdao-prefix> + <clean UUID>`.

danxmoran reviewed (commented):
Didn't take a close look at the tests

fboulnois commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DataDeletionStep.java:
```diff
@@ -0,0 +1,86 @@
+        this.datasetService = datasetService;
+    }
+
+    private String suffix(FlightContext context) {
```
This method is used in a few places, perhaps pull it out into utils?

thathert replied:
Good point, was wondering the same.

fboulnois reviewed (approved):
I echo some of the earlier suggestions. Other minor addition:

Commit:
Update src/main/resources/data-repository-openapi.yaml

Co-Authored-By: Daniel Moran <danmoran@broadinstitute.org>

Commit:
PR comments

Commit:
Merge branch 'jh-fb-DR-479-soft-deletes-api' of github.com:DataBiosphere/jade-data-repo into jh-fb-DR-479-soft-deletes-api

danxmoran commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DataDeletionStep.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public StepResult undoStep(FlightContext context) {
+        // The do step is atomic, either it worked or it didn't. There is no undo.
```
:yoda:

danxmoran commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DropExternalTablesStep.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private String suffix(FlightContext context) {
+        return context.getFlightId().replace('-', '_');
```
Could use `DataDeletionUtils` here? For this method and for `request` below, I think

thathert replied:
Yep, missed it on refactor.

danxmoran commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DropExternalTablesStep.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    public StepResult undoStep(FlightContext context) {
+        // TODO: need a human to intervene -- the soft delete worked but we couldn't clean up tables
```
A thought, not for this ticket: BQ supports [per-table TTLs](https://cloud.google.com/bigquery/docs/managing-tables#updating_a_tables_expiration_time), it could be useful for the repo to use that setting on the temp tables it creates to help mitigate situations like this.

danxmoran commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+        Schema schema = Schema.of(Field.of(PDAO_ROW_ID_COLUMN, LegacySQLTypeName.STRING));
+        ExternalTableDefinition tableDef = ExternalTableDefinition.of(path, schema, FormatOptions.csv());
+        TableInfo tableInfo = TableInfo.of(tableId, tableDef);
+        bigQueryProject.getBigQuery().create(tableInfo);
```
Should we try to catch & inspect any exceptions raised here, so we can convert errors from a wildcard that matches 0 files into a 404 error? We do [this hack](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java#L455-L457) when using a load job, not sure what the equivalent would be for an external table definition.

thathert replied:
Thanks for the pointer, yes I'll add this here and also move the missing non-wildcard file check here like you mention in the previous comment.

danxmoran commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+                .add("project", bigQueryProject.getProjectId())
+                .add("dataset", prefixName(dataset.getName()))
+                .add("softDeleteTable", softDeleteTableNameLookup.get(tableName))
+                .add("rowId", PDAO_ROW_ID_COLUMN)
```
This breaks the convention we use elsewhere in the file (using `+` to hard-code static constants into the SQL templates up-front). I like this way a _lot_ more in terms of readability, but thought I should point it out.

thathert replied:
👍I wanted to try it out to see the difference -- I find this way more readable as well.

danxmoran commented on src/test/java/bio/terra/service/dataset/DatasetIntegrationTest.java:
```diff
@@ -195,4 +216,118 @@ public void testAssetCreationUndo() throws Exception {
+            .specType(DataDeletionRequest.SpecTypeEnum.GCSFILE);
+    }
+
+    // TODO: wildcard test, validation tests, empty file, missing file, fault insertion on creating ext tables
```
Wildcard test is done, yes? Though could use a 404 wildcard case.

marikomedlock replied:
I think a test that soft deletes the same rowid twice would be good.

danxmoran commented on src/test/java/bio/terra/service/dataset/DatasetIntegrationTest.java:
```diff
@@ -195,4 +216,118 @@ public void testAssetCreationUndo() throws Exception {
+        ingestResponse = dataRepoFixtures.ingestJsonData(steward(), datasetId, ingestRequest);
+        assertThat("correct sample row count", ingestResponse.getRowCount(), equalTo(7L));
+        return datasetId;
+    }
```
I think variations of all these helpers existing in the `BigQueryPdao` connected test. Would anything here be easier / more reliable from testing at that level instead of at the full integration level? I guess the lines have already blurred pretty badly between the two kinds of tests...

danxmoran commented on src/main/java/bio/terra/service/dataset/flight/datadelete/CreateExternalTablesStep.java:
```diff
@@ -0,0 +1,115 @@
+                // There's not an easy way to check this with the SDK. Let BQ handle catching this on its side.
+                return true;
+            } else {
+                return GcsPdao.getBlobFromGsPath(storage, path).exists();
```
If we're ok letting BQ be the failure point for the wildcard case, are we also ok letting it be the same for the single-file case? It'd save us up-front work / API calls.

ddietterich replied:
+1

danxmoran reviewed (approved):
Would be ok merging as-is, but left some thoughts

ddietterich commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DataDeletionStep.java:
```diff
@@ -0,0 +1,56 @@
+        bigQueryPdao.validateDeleteRequest(dataset, dataDeletionRequest.getTables(), suffix);
+        bigQueryPdao.applySoftDeletes(dataset, dataDeletionRequest.getTables(), suffix);
+
+        // TODO: this can be more informative, something like # rows deleted per table, or mismatched row ids
```
Is this extra information required now? If so, this should probably have a Jira ticket reference

thathert replied:
I will make a ticket

- [x] made ticket and linked it here.

thathert replied:
https://broadworkbench.atlassian.net/browse/DR-911

ddietterich commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+            // shortcut out early, no use wasting more compute
+            if (numMismatched > 0) {
+                throw new MismatchedRowIdException(
+                    String.format("Could not match %s row ids for table %s", numMismatched, tableName));
```
@danxmoran Is this sufficient error information for you?

ddietterich commented on src/main/resources/data-repository-openapi.yaml:
```diff
@@ -2658,6 +2692,56 @@ definitions:
+      gcsFileSpec:
+        $ref: '#/definitions/DataDeletionGcsFileModel'
+
+  DataDeletionGcsFileModel:
```
This is not yet implemented, right?  Usually we don't put APIs in here on spec.

thathert replied:
DanM had mentioned these names are confusing, and I think this is confirmation of it :)

This model is used, it is the "object that contains a path property with a gs URI that points to a file that contains row ids for a specific table to be soft deleted". I'm having trouble finding a concise name.

Maybe switching to `TabularDataDeletion` for all of these prefixes, and instead of calling this a `GcsFile` it can be a `GcsUri` to avoid confusion with file deletions? That would leave this: `TabularDataDeletionGcsUriModel`

ddietterich reviewed (changes requested):
I'd really like to see the lock interfaces changed to use ids instead of names, one way or the other.

I still don't get the wildcard use case. I checked the brief design doc and didn't find any info there.

I'd also like to confirm that the error responses are sufficient for Monster. We could do a bunch more work and make very informative ones, but if they are not needed, I would skip them for now.

thathert commented:
> I'd really like to see the lock interfaces changed to use ids instead of names, one way or the other.

@marikomedlock what do you think about this?

Sometimes a request includes a dataset name, sometimes it just includes a dataset id. A previous iteration of this branch added a `Step` that fetched a dataset by the dataset id and then put its name in the workingMap. I had updated the Lock steps to be able to grab the dataset name from the workingMap instead of receiving it as a constructor param.

I figure that there will be cases where there is a database blip that could prevent DAO lookups, so instead of doing those lookups in the Flight constructor we should perform them inside of Steps that can decide to retry, etc. I switched off of this because it felt unnecessary to do the extra lookup -- intuitively we should be able to lock based on either of those unique columns.

@ddietterich aside from the admitted awkwardness of the method I implemented to handle the lookup, what are the other cons of this approach that you see?

ddietterich commented:
I would prefer not using names. As I understand it, there are two locking cases:
- create, where we create and lock in one operation.
- everything else, where we always have a dataset id; but we don't always have a name.
I believe that is the pattern Mariko implemented for both dataset and snapshot.

Am I wrong about that? If not, it seems better to have just the id version of locking since it is universal.

Typically in the DAOs, we have two lookup methods: one by name and one by id. Another alternative is to have two lock methods as well. Then the right one can be used in the right context.

If it would be useful, I think it would be straightforward for lock() to return the object it locked. On the lock update, we can use the key holder to retrieve the data.

A less strong reason, but a concern of mine is that the current design of Folder Manager is that it holds the names of datasets and snapshots. The less we have name dependencies, the easier that eventual integration will be.

I appreciate your line of reasoning about putting dataset retrieval into the steps and be able to retry it.

marikomedlock commented:
In the create and lock case, we have to lock on the name because there's no common id written to PostGres, yet.

In the lock existing record case, I think we should just lock on whichever one allows us to avoid separate id <-> name conversions. So if the request specifies a Dataset by its id, then I think we should lock on the id. That's the case you have here, so I think you can change the lock() method to use the id instead. In the future, if we need to lock a dataset/snapshot and the request only specifies its name, then that would be the time to add a lockOnName() method. I'm not sure off the top of my head if we have a case like that, yet, but my preference is just to avoid unnecessary id/name conversions.

marikomedlock commented on src/main/java/bio/terra/service/dataset/flight/LockDatasetBaseStep.java:
```diff
@@ -0,0 +1,52 @@
+        }
+        String datasetId = context.getInputParameters().get(JobMapKeys.DATASET_ID.getKeyName(), String.class);
+        if (datasetId == null) {
+            throw new InvalidLockArgumentException("Must pass dataset name to step constructor or provide datasetId " +
```
Given the comments in the main thread on this PR, I don't think you need this BaseStep class anymore. In the future, if we add a lockByName() method to dataset/snapshot dao, then I think we can just modify the existing LockDatasetStep to use the name if the id is not in the context.

But if you do still throw this exception, I think it would be helpful to rename it to something that doesn't include "lock". This is more of a bad context state and the lock in the exception name was confusing to me.

marikomedlock commented on src/main/java/bio/terra/service/dataset/flight/LockDatasetBaseStep.java:
```diff
@@ -0,0 +1,52 @@
+            throw new InvalidLockArgumentException("Must pass dataset name to step constructor or provide datasetId " +
+                "in the flight input params");
+        }
+        Dataset dataset = datasetDao.retrieve(UUID.fromString(datasetId));
```
This is the id<->name conversion that I'd like to avoid that I mentioned in my comment in the main thread.

marikomedlock commented on src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java:
```diff
@@ -2,44 +2,49 @@
     public StepResult doStep(FlightContext context) {
         try {
-            datasetDao.lock(datasetName, context.getFlightId());
+            getDatasetDao().lock(getDatasetName(context), context.getFlightId());
```
Do you need a getter for this property? If you just want subclasses to access it, can you make it protected instead?

marikomedlock commented on src/main/java/bio/terra/service/dataset/flight/datadelete/CreateExternalTablesStep.java:
```diff
@@ -0,0 +1,115 @@
+        String suffix = getSuffix(context);
+
+        for (DataDeletionTableModel table : getRequest(context).getTables()) {
+            bigQueryPdao.deleteSoftDeleteExternalTable(dataset, table.getTableName(), suffix);
```
How about adding a try/catch around this line for a BigQueryException, so that we cleanup the best we can, even if one fails?

marikomedlock commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DropExternalTablesStep.java:
```diff
@@ -0,0 +1,62 @@
+            .get(JobMapKeys.REQUEST.getKeyName(), DataDeletionRequest.class);
+    }
+
+    private Dataset dataset(FlightContext context) {
```
I think this method is doing the same thing as the DataDeletionUtils method. Maybe you meant to replace it?

thathert replied:
Ah I thought I factored this out. Thanks, good catch!

marikomedlock commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DropExternalTablesStep.java:
```diff
@@ -0,0 +1,62 @@
+        DataDeletionRequest dataDeletionRequest = request(context);
+
+        dataDeletionRequest.getTables().forEach(table ->
+            bigQueryPdao.deleteSoftDeleteExternalTable(dataset, table.getTableName(), suffix));
```
Same comment as above, perhaps add a try/catch BigQueryException around each delete so we clean up as much as possible? And then log the problematic tables, for future reference by the human cleaner upper.

marikomedlock commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+        return bigQueryProject.deleteTable(prefixName(dataset.getName()), extTableName);
+    }
+
+    private static final String insertSoftDeleteTemplate =
+        "INSERT INTO `<project>.<dataset>.<softDeleteTable>` " +
+        "SELECT DISTINCT <rowId> FROM `<project>.<dataset>.<softDeleteExtTable>`";
```
I really like this templating. Very easy to follow your SQL logic, thanks!

marikomedlock commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+                                 String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+
+        // we want this soft delete operation to be atomic, we we will combine all of the inserts into one statement
+        // that we send to bigquery.
```
Could you elaborate more on the "atomic" part of the comment please? I don't think running a SQL script or multiple statements separated by semicolons makes any additional guarantees when compared with running the statements individually, though I'm not familiar with BigQuery's behavior specifically. Do you mean that you just want to make a single BQ network request?

thathert replied:
This combines them all into one BigQuery `Job`. I was hoping that this would treat them as a transaction that succeeds or fails atomically. On further investigation it looks like [BigQuery creates a child job for each statement](https://cloud.google.com/bigquery/docs/reference/standard-sql/scripting#bigquery-scripting), so my hopes are dashed! I'll update the comment and make a ticket to revisit this. I think we need something along the lines of a soft delete tag to be able to get transaction like behavior in the case of a failure in one of these child jobs.

marikomedlock replied:
Making a separate ticket to address this sounds good. Especially since soft-deleting the same row twice is not allowed, so users can't just keep executing the same request until it all succeeds.

thathert replied:
https://broadworkbench.atlassian.net/browse/DR-910

marikomedlock commented on src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java:
```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+     * Note that since this is joining against the live view, an attempt to soft delete a rowId that has already been
+     * soft deleted will result in a mismatch.
+     */
+    private static final String validateSoftDeleteTemplate =
+        "SELECT COUNT(E.<rowId>) FROM `<project>.<dataset>.<softDeleteExtTable>` E " +
+        "LEFT JOIN `<project>.<dataset>.<liveView>` V USING (<rowId>) " +
+        "WHERE V.<rowId> IS NULL";
```
Joining against the live view instead of the underlying table means that soft deleting the same rowid twice will throw an error, correct? Is that your intention?

thathert replied:
Yep :)

marikomedlock reviewed (commented):
I'd appreciate answers to a couple of questions in the BigQueryPdao class. Otherwise looks good to me.

Regarding the locking on name/id question, I'd be happy to put in a quick PR that changes the Lock step to use the id so that you can call it here instead of writing a new step with an id->name translation. Let me know if you'd prefer that to just including any changes in your PR directly.

Commit:
remove file checking from step

Commit:
switch to datadeletionutils

Commit:
switch to locking by id

Commit:
address pr comments, push row id file checking down to pdao layer
